source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -437,7 +437,8 @@ void Filter::UpstreamCallbacks::onEvent(Network::ConnectionEvent event) {
 
 void Filter::UpstreamCallbacks::onAboveWriteBufferHighWatermark() {
   // TCP Tunneling may call on high/low watermark multiple times.
-  ASSERT(parent_->config_->tunnelingConfigHelper() || !on_high_watermark_called_);
+  ASSERT((parent_ != nullptr && parent_->config_->tunnelingConfigHelper()) ||
```

Comment:
in the case when we enter draining state `parent_` will be set to nullptr. In draining state if code invokes this callback is will trigger ASSERT failure in non release mode. @ggreenway would this be correct logic to assert fail if callback is invoked in draining state? or should we instead have ASSERT statement like:
```
ASSERT(parent == nullptr || parent_->config_->tunnelingConfigHelper() || !on_high_watermark_called_);
```